packages/contracts/src/contracts/current/protocol/Exchange/MixinSignatureValidator.sol

```diff
@@ -0,0 +1,138 @@
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.4.19;
+
+import "./mixins/MSignatureValidator.sol";
+import "./ISigner.sol";
+
+/// @dev Provides MSignatureValidator
+contract MixinSignatureValidator is
+    MSignatureValidator
+{
+    enum SignatureType {
+        Illegal, // Default value
+        Invalid,
+        Caller,
+        Ecrecover,
+        EIP712,
+        Contract
+    }
+  
+    function isValidSignature(
+        bytes32 hash,
+        address signer,
+        bytes signature)
+        public view
+        returns (bool isValid)
+    {
+        // TODO: Domain separation: make hash depend on role. (Taker sig should not be valid as maker sig, etc.)
+        
+        require(signature.length >= 1);
+        SignatureType signatureType = SignatureType(uint8(signature[0]));
+        
+        // Variables are not scoped in Solidity
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+        
+        // Always illegal signature
+        // This is always an implicit option, since a signer can create a
+        // signature array with invalid type or length. We may as well make
+        // it an explicit option. This aids testing and analysis. It is
+        // also the initialization value for the enum type.
+        if (signatureType == SignatureType.Illegal) {
+            revert();
+        
+        // Always invalid signature
+        // Like Illegal, this is always implicitely available and therefore
+        // offered dxplicitely. It can be implicitely creates by providing
+        // a validly formatted but incorrect signature.
+        } else if (signatureType == SignatureType.Invalid) {
+            require(signature.length == 1);
+            isValid = false;
+            return isValid;
+        
+        // Implicitly signed by caller
+        // The signer has initiated the call. In the case of non-contract
+        // accounts it means the transaction itself was signed.
+        // Example: lets say for a particular operation three signatures
+        // A, B  are required. To submit the transaction, A and B can give
+        // a signatue to C, who can then submit the transaction using
+        // `Caller` for his own signature. Or A and C can sign and B can
+        // submit using `Caller`. Having `Caller` allows this flexibility.
+        } else if (signatureType == SignatureType.Caller) {
+            require(signature.length == 1);
+            isValid = signer == msg.sender;
+            return isValid;
+        
+        // Signed using web3.eth_sign
+        } else if (signatureType == SignatureType.Ecrecover) {
+            require(signature.length == 66);
+            v = uint8(signature[1]);
+            r = get32(signature, 2);
+            s = get32(signature, 34);
+            recovered = ecrecover(
+                keccak256("\x19Ethereum Signed Message:\n32", hash),
```

Comment:
Let's add support for current Trezors which do not use ASCII characters for the message length encoding: https://github.com/ethereum/go-ethereum/issues/14794.

Reply:
I'm glad you're adding support for this, but can't give a thumbs up because it's obnoxious it's different in the first place.

Reply:
Actually, I need to side with the Trezor guys here. The way `eth_sign` does it is a security hole. It allows two different messages to have the same hash.

Reply:
Oh, agreed. varInt is much better than an ASCII representation. It's just so strange there doesn't appear to be an EIP around it and everyone (other than Trezor) just supports both ways like it's no big deal.

Reply:
We are really trying to push the community towards https://github.com/ethereum/EIPs/pull/712